CANBridge: Add support for Linux and macOS

qwertychouskie opened the pull request:
The @unofficial-rev-port team has been hard at work making the core technologies used by the REV Hardware Client work on all 3 major platforms (Windows, macOS, Linux).  This PR contains changes needed for CANBridge to compile and run on macOS and Linux in addition to Windows; and the GitHub Actions changes needed to upload binaries for all 3 OSes for use by the node-can-bridge build process to pull the correct libraries for the 3 platforms (see https://github.com/REVrobotics/node-can-bridge/pull/21).

Please note that this PR is still a work-in-progress; while we have successfully ran a local build of the REV Hardware Client that was able to load and initialize the CAN backend, we are still in the process of testing with real hardware to ensure full functionality (and no regressions) on all 3 OSes.  We are opening this PR in the current state in order to gather feedback and make sure we do the changes in a way acceptable to having the PR merged upstream (once completed, of course).

willtoth committed:
Start of SocketCAN implementation

willtoth committed:
Simple method to list available drivers

willtoth committed:
Update README.md

Moose1301 committed:
Made SerialDevice complie when linux

Note you might need to complie wpilib your self to complie this yourself
this is due to wpilib not compling for all versions of linux in the maven repo

Moose1301 committed:
Fix compling on windows

qwertychouskie committed:
CI: Initial build config for macOS and Linux for both x86-64 and arm64

qwertychouskie committed:
Build: Disable warnings-as-errors

There's a lot of warnings that apparently aren't showing up on the Windows build

garrettsummerfi3ld committed:
Add linker argument for IOKit framework on macOS universal builds

Fixes broken CANBridge builds on any macOS platform.

TBHGodPro committed:
update build.yml for macosarm64

TBHGodPro committed:
patch build gh-workflow for macosarm64

garrettsummerfi3ld committed:
Update artifact workflows, added macOS artifact creation

Updated workflows to better handle artifact uploads. This includes better cleanup of no longer dumping all the possible platforms into the main directory of the CI job, as well as for each platform type, specific artifacts for each platform is made to lower artifact sizes and to be specific to the correct platform.

Added other adjustments for platform types across the workflow for future ARM-specific builds

Added macOS artifact creation.

garrettsummerfi3ld committed:
Add Arm32, Arm64 builds, upload Linux artifacts

Added builds to utilize Linux ARM 32 and 64 bit versions.

Added artifact uploads for Linux versions.

garrettsummerfi3ld committed:
[ci-skip] Add name to CI jobs

garrettsummerfi3ld committed:
Disable fail-fast, install 7z during Docker builds

Disabled fail-fast to allow for all other jobs to continue running without interruption, especially if another job fails inside the matrix.

Added step to install 7Zip for the Docker CI. This is just to zip and archive up packages without issues.

garrettsummerfi3ld committed:
Separate step for 7zip installation

garrettsummerfi3ld committed:
Enable verbose logging during build outputs on ARM builds

garrettsummerfi3ld committed:
Finalize ARM builds and artifact packaging

garrettsummerfi3ld committed:
Fix incorrect directories for creating artifacts

garrettsummerfi3ld committed:
Split steps apart for WPILib HAL archiving, update release.yml

Steps for unarchiving and archiving HALs are split up to add verbosity to the CI steps, if a specific step fails, we are able to diagnose more quickly.

Updated release CI steps, using updated Actions and cleaner methods to check if the build CI completed.

garrettsummerfi3ld committed:
Update dependent actions

Updated dependent actions due to deprecation notices.

garrettsummerfi3ld committed:
Merge pull request #3 from unofficial-rev-port/ci/artifact-uploads

Update artifact uploading workflows, add ARM workflows, update release workflows

garrettsummerfi3ld committed:
[ci skip] Updating release workflow to properly check for workflow releases

garrettsummerfi3ld committed:
[ci skip] Update tags search release workflow

garrettsummerfi3ld committed:
Remove tag name

garrettsummerfi3ld committed:
Add GITHUB_TOKEN to release workflow and permissions to write releases

garrettsummerfi3ld committed:
Fix broken check waiting for build to finish.

garrettsummerfi3ld committed:
Shorten artifact names, skip unpacking for releases

garrettsummerfi3ld committed:
[ci skip] Remove SHA on ARM builds for artifacts

garrettsummerfi3ld committed:
chore: Update WPILib HAL headers in release workflow

This updates workflows to produce headers for CANBridge separately from the built binaries.

garrettsummerfi3ld committed:
chore: Fix missing directory in build artifacts

garrettsummerfi3ld committed:
chore: Fix missing directory for build artifacts (again)

garrettsummerfi3ld committed:
Remove macOS ARM workflow

Built binaries for macOS are universal for ARM and x86.

garrettsummerfi3ld committed:
fix: Un-nest the headers artifact

The headers artifact was nested by mistake, this is now resolved.

garrettsummerfi3ld committed:
fix: Actually copy everything in `include` for `headers.zip`

garrettsummerfi3ld committed:
Update `README.md`

Updated compile instructions and formatting.

qwertychouskie committed:
Attempt to fix runtime crash

qwertychouskie committed:
(hopefully) Fixup Windows build

qwertychouskie committed:
v2.3.5

qwertychouskie committed:
Serial: More debugging

qwertychouskie committed:
Attempt to fix C++ exceptions on macOS

qwertychouskie committed:
Attempt to fix C++ exceptions on macOS part 2

qwertychouskie committed:
Attempt to fix C++ exceptions on macOS part 3

qwertychouskie committed:
Attempt to fix C++ exceptions on macOS part 4: undo da changes

The linker flag wasn't needed apparently

garrettsummerfi3ld commented:
Other notes to be made, we have added CI to this to validate not only the already existing workflows, but also new builds to support Linux (ARM32/64, and x86) and macOS.

NoahAndrews commented:
Thank you for contributing! I expect that this work will be the hardest part of bringing Mac and Linux support to the REV Hardware Client, so if you're successful here, that increases the odds of that happening significantly.

I see that you've started with porting the serial driver. Unfortunately, the serial driver only works with SPARK MAX firmware 1.4 and earlier, and the REV Hardware Client doesn't use it at all (it has its own minimal implementation of the serial protocol that is only capable of putting the device in dfu mode).

For Linux support, see the [SocketCAN branch](https://github.com/REVrobotics/CANBridge/tree/SocketCAN?tab=readme-ov-file#linux). The Linux kernel actually contains a gs_usb driver that is compatible with our USB CAN devices, so on Linux we can take advantage of that, and use Linux's SocketCAN interface instead of writing low-level USB code. I'm not sure what shape that branch is in, but I would recommend starting a new branch based on that one, merging `main` into it, porting over your CI and build changes, and going from there. I would also be interested in merging a PR that deletes the serial driver entirely.

The big lift will be writing a candle/gs_usb driver for Mac. Unfortunately, that protocol is not documented anywhere that I've seen, so if you want to tackle that, you're probably going to have to learn it by reading the `candlelib` code in this repository (originally published [here](https://github.com/HubertD/candle_dll)) and [kernel gs_usb driver](https://github.com/torvalds/linux/blob/master/drivers/net/can/usb/gs_usb.c) source code. There's also a [python implementation](https://github.com/jxltom/gs_usb).

Please let me know if you have any questions! Once you have some code that will work with modern REV firmware, I'll give it a more thorough review.

garrettsummerfi3ld committed:
Merge branch 'main' into SocketCAN

Brings SocketCAN from upstream back to origin with already existing changes from main for workflows and other changes.

garrettsummerfi3ld committed:
Update SocketCANDevice class implementation

Some of the functions and implementations were not readily available nor created. This now resolves those issues.

garrettsummerfi3ld committed:
chore: Update googletest dependency

The googletest dependency was using an older version from the WPILib Artifactory, this was causing issues with builds on macOS where `osxuniversal` platform type was not available, and split between `osxarm64` and `osxx86-64`.

garrettsummerfi3ld committed:
feat: Implement SocketCANDevice class

This commit adds the implementation for the SocketCANDevice class, which was previously missing some functions and implementations. Now all required functions are available and the issues have been resolved.

garrettsummerfi3ld committed:
Implement SocketCANDeviceThread in SocketCANDevice class

This was missing the previous commit.

Change compiler options to have SocketCANThread actually compile instead of not.

garrettsummerfi3ld committed:
refactor: Update include path for CanData header in SocketCANThread.h

Mirroring SerialDeviceThread.h as it was changed years ago.

garrettsummerfi3ld committed:
Fix compiler issues and broken functions

Adding more compiler args to prevent other platforms that are not able to compile some platforms to compile files, as it would just error and fail the build.

Fixing broken functions such as `SocketCANDevice::GetID()` as it was not returning a string. This is now not implemented but it returns an `int` which is correct for the function.

garrettsummerfi3ld committed:
Fix SocketCANDevice compilation and broken functions

Fix SocketCANDevice from being broken on compiles, from broken constructors to missing includes.

garrettsummerfi3ld committed:
Implement SocketCANDriver

This is largely untested, however this should bring the functionality of the SocketCAN work somewhat back into play.

garrettsummerfi3ld committed:
Rewrite SocketCANThread header

This tears out the majority of what was [candle_dll](https://github.com/HubertD/candle_dll/tree/master) for the Linux driver and implemented the Linux [SocketCAN](https://docs.kernel.org/networking/can.html) driver.

Preliminary testing shows that this does functionally work with a single SPARK MAX motor controller, however further testing may be required.

garrettsummerfi3ld committed:
Update README.md

Updated section for Linux support.

garrettsummerfi3ld committed:
Remove dangling serial driver

Plans to fully delete the serial driver are underway, as this is not supported or even functioning with current firmware.

garrettsummerfi3ld committed:
Change hardcoded name for SocketCANDevice

SocketCAN devices aren't exclusively, changing it to something more generic.

garrettsummerfi3ld committed:
Merge pull request #5 from unofficial-rev-port/SocketCAN

Add SocketCAN functionality

garrettsummerfi3ld committed:
Update GoogleTest harness

The current implementation of the tests were accounting for only Linux platforms for SocketCAN, this has been changed to allow for all currently supported platforms. This also removes conditionals that prevented any GoogleTests from running.

NoahAndrews commented:
Note to self: Test timestamp resolution on new platforms before merging

garrettsummerfi3ld committed:
Merge remote-tracking branch 'upstream/main'

LandryNorris commented:
We're adding support for RTR messages (Remote Transmission Request) in #36. The way an RTR is indicated is unfortunately not very standardized. WPILib uses one of the reserved bits (0x40000000) of the ARB ID. Since there's no standard way of specifying this (candlelib uses 0x80000000, for example), each driver has to handle the translation. See the diff in #36 for how Candlelib will handle this translation.

garrettsummerfi3ld commented:
This branch should be enough to _compile_ CANBridge but has none of the function of CANBridge.

I am working on adding a driver since the driver being used only works on Windows, but at the moment it is a bit stalled. I am picking it back up again and looking to have this completed soon.

jan-raghavan commented:
What is the status here? I mean I still there are lots of conflicts to be resolved. I am looking to use this rev hardware client for my MAC
